lang/java/reef-utils/src/main/java/org/apache/reef/util/ComplexCondition.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.reef.util;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Manages Java lock and condition objects to create a simplified
+ * condition variable interface.
+ */
+public final class ComplexCondition {
```

Comment:
If we make `ComplexCondition` implement `AutoCloseable` and call `.releaseLock()` on `.close()`, we can use it in `try` block, e.g.
```java
    try (final ComplexCondition call = getSleeper(identifier).takeLock()) {
      LOG.log(Level.FINER, "Waking caller sleeping on identifier [{0}]", identifier);
      call.signalCondition();
    }
```
(assuming `.takeLock()` returns `this`)

P.S. not 100% sure about it, though, as it is a noticeable departure from the Java `Lock` interface.

P.P.S. if we decide not to implement `AutoCloseable`, we should probably bring our API closer to the standard Java naming scheme - e.g. use `.lock()` and `.unlock()` instead of `.takeLock()` and `.releaseLock()` etc.